server/cp/output/formatter/ninjs_formatter_2.py

```diff
@@ -0,0 +1,692 @@
+    Added *evolvedfrom* field to ninjs output.
+
+"""
+
+
+import json
+import superdesk
+import logging
+import re
+from typing import Tuple
+
+from flask import current_app as app
+from eve.utils import config
+from superdesk.publish.formatters import Formatter
+from superdesk.errors import FormatterError
+from superdesk.metadata.item import ITEM_TYPE, CONTENT_TYPE, EMBARGO, GUID_FIELD, ASSOCIATIONS
+from superdesk.metadata.packages import RESIDREF, GROUP_ID, GROUPS, ROOT_GROUP, REFS
+from superdesk.utils import json_serialize_datetime_objectId
+from superdesk.media.renditions import get_renditions_spec
+from superdesk.vocabularies import is_related_content
+from apps.archive.common import get_utc_schedule
+from superdesk import text_utils
+from superdesk.attachments import get_attachment_public_url
+from collections import OrderedDict
+
+logger = logging.getLogger(__name__)
+# this regex match the way custom media fields are put in associations (i.e. how the key
+# is generated). This is legacy, and can hardly be changed without risking to break
+# instances in production.
+MEDIA_FIELD_RE = re.compile(r"(?P<field_id>\S+)--(?P<version>\d+)$")
+EXTRA_ITEMS = "extra_items"
+
+SCHEME_MAP = {
+    "sig": "http://cv.iptc.org/newscodes/signal/",
+}
+
+
+def filter_empty_vals(data):
+    """Filter out `None` values from a given dict."""
+    return dict(filter(lambda x: x[1], data.items()))
+
+
+def get_locale_name(item, language):
+    try:
+        return item["translations"]["name"][language]
+    except (KeyError, TypeError):
+        return item.get("name", "")
+
+
+def format_cv_item(item, language):
+    """Format item from controlled vocabulary for output."""
+    if item.get("scheme") == "subject":
+
+        return filter_empty_vals(
+        {"code": item.get("qcode"), "name": get_locale_name(item, language), "scheme": "http://cv.iptc.org/newscodes/mediatopic/"}
+    )
+    else:
+
+        return filter_empty_vals(
+            {"code": item.get("qcode"), "name": get_locale_name(item, language), "scheme": item.get("scheme")}
+        )
+
+
+class NINJSFormatter_2(Formatter):
+    """
+    The schema we use for the ninjs format is an extension
+    of `the standard ninjs schema <http://www.iptc.org/std/ninjs/ninjs-schema_1.1.json>`_.
+
+    *Changes from ninjs schema*:
+
+    * ``uri`` was replaced by ``guid``: ``uri`` should be the resource identifier on the web
+        but since the item was not published yet it can't be determined at this point
+    * added ``priority`` field
+    * added ``service`` field
+    * added ``slugline`` field
+    * added ``keywords`` field
+    * added ``evolvedfrom`` field
+    * added ``source`` field
+
+    Associations dictionary may contain entire items like
+    in `ninjs example <http://dev.iptc.org/ninjs-Examples-3>`_ or just the item ``guid``
+    and ``type``. In the latest case the items are sent separately before the package item.
+    """
+
+    type = "ninjs"
+    name = "NINJS"
```

Comment:
would be good to have a custom name/type for this unless you want to replace existing formatter in current setup

Reply:
resolved with latest update